File: JAVA/Pertemuan 05/src/id/ac/uph/pbo/pertemuan05/KonstruktorJava.java
package id.ac.uph.pbo.pertemuan05;

public class KonstruktorJava {

    // Demo nilai default Java
    static void demoNilaiDefault() {
        System.out.println("\n─── NILAI DEFAULT TIPE DATA DI JAVA ─────────");
        // Deklarasi tanpa inisialisasi → Java set nilai default
        // (hanya untuk atribut class, BUKAN variabel lokal)
        System.out.println("  int      → 0");
        System.out.println("  double   → 0.0");
        System.out.println("  boolean  → false");
        System.out.println("  char     → '\\u0000' (null char)");
        System.out.println("  Object   → null");
        System.out.println("  String   → null");
    }

    public static void main(String[] args) {
        System.out.println("╔══════════════════════════════════════════════╗");
        System.out.println("║   PERTEMUAN 5 – Konstruktor Default & Non-Default ║");
        System.out.println("╚══════════════════════════════════════════════╝");

        // ─── Demo MataKuliah ──────────────────────────────────────────────
        System.out.println("\n═══ DEMO: Konstruktor MataKuliah ═══");
        System.out.println("Membuat objek dengan konstruktor berbeda:\n");
        MataKuliah mk1 = new MataKuliah();                          // default
        MataKuliah mk2 = new MataKuliah("SYS07406", "PBO");        // 2-param
        MataKuliah mk3 = new MataKuliah("SYS07301", "Basis Data", 3); // 3-param
        MataKuliah mk4 = new MataKuliah("SYS07100", "Kalkulus", 3, "Wajib", 60.0); // lengkap
        MataKuliah mk5 = new MataKuliah(mk4);                       // copy constructor

        System.out.println("\nDaftar Mata Kuliah:");
        System.out.printf("  %-12s | %-40s | SKS | Tipe     | Min%n", "Kode", "Nama");
        System.out.println("  " + "─".repeat(80));
        mk1.tampilkan();
        mk2.tampilkan();
        mk3.tampilkan();
        mk4.tampilkan();
        mk5.tampilkan();

        // ─── Demo Mahasiswa ─────────────────────────────────────────────
        System.out.println("\n═══ DEMO: Konstruktor Mahasiswa Overloading ═══");
        Mahasiswa m1 = new Mahasiswa("25SI001", "Budi");
        Mahasiswa m2 = new Mahasiswa("25SI002", "Sari", "Teknik Informatika");
        Mahasiswa m3 = new Mahasiswa("25SI003", "Andi", "Sistem Informasi", 2024, 3.85);

        m1.setIpk(3.50);

        System.out.printf("  %-12s | %-20s | %-20s | Tahun | IPK%n", "NIM", "Nama", "Prodi");
        System.out.println("  " + "─".repeat(80));
        m1.tampilkan();
        m2.tampilkan();
        m3.tampilkan();

        // ─── Nilai default ────────────────────────────────────────────────
        demoNilaiDefault();

        System.out.println("\n─── KESIMPULAN ──────────────────────────────");
        System.out.println("  Konstruktor Default     : dipanggil tanpa argumen");
        System.out.println("  Konstruktor Non-Default : minimal 1 argumen");
        System.out.println("  Overloading Konstruktor : banyak konstruktor, parameter beda");
        System.out.println("  this(...)               : memanggil konstruktor lain di class sama");
        System.out.println("  Copy Constructor        : membuat salinan objek lain");
    }
}

File: PYTHON/Pertemuan 05/KonstruktorJava.py
import sys
sys.stdout.reconfigure(encoding='utf-8')
from MataKuliah import MataKuliah
from Mahasiswa import Mahasiswa

# Demo nilai default (disimulasikan karena Python tidak punya tipe primitif dengan default bawaan otomatis selain tipe default argumen)
def demo_nilai_default():
    print("\n─── NILAI DEFAULT TIPE DATA DI JAVA vs PYTHON ─────────")
    print("Di Python, atribut kelas tidak memiliki nilai default otomatis kecuali diinisialisasi secara eksplisit, biasanya di __init__")
    print("  int      → biasanya diinisialisasi 0")
    print("  float    → biasanya diinisialisasi 0.0")
    print("  bool     → biasanya diinisialisasi False")
    print("  str      → biasanya diinisialisasi '' atau None")
    print("  objek    → biasanya diinisialisasi None")

def main():
    print("╔══════════════════════════════════════════════╗")
    print("║   PERTEMUAN 5 – Konstruktor Default & Non-Default ║")
    print("╚══════════════════════════════════════════════╝")

    # ─── Demo MataKuliah ──────────────────────────────────────────────
    print("\n═══ DEMO: Konstruktor MataKuliah ═══")
    print("Membuat objek dengan konstruktor (argumen default) berbeda:\n")
    mk1 = MataKuliah()                                           # default
    mk2 = MataKuliah(kode="SYS07406", nama="PBO")                # 2-param
    mk3 = MataKuliah(kode="SYS07301", nama="Basis Data", sks=3)  # 3-param
    mk4 = MataKuliah(kode="SYS07100", nama="Kalkulus", sks=3, tipe="Wajib", nilai_min=60.0) # lengkap
    mk5 = MataKuliah(lain=mk4)                                   # copy constructor

    print("\nDaftar Mata Kuliah:")
    print(f"  {'Kode':<12} | {'Nama':<40} | SKS | Tipe     | Min")
    print("  " + "─" * 80)
    mk1.tampilkan()
    mk2.tampilkan()
    mk3.tampilkan()
    mk4.tampilkan()
    mk5.tampilkan()

    # ─── Demo Mahasiswa ─────────────────────────────────────────────
    print("\n═══ DEMO: Konstruktor Mahasiswa Overloading ═══")
    m1 = Mahasiswa("25SI001", "Budi")
    m2 = Mahasiswa("25SI002", "Sari", program_studi="Teknik Informatika")
    m3 = Mahasiswa("25SI003", "Andi", program_studi="Sistem Informasi", tahun_masuk=2024, ipk=3.85)

    m1.set_ipk(3.50)

    print(f"  {'NIM':<12} | {'Nama':<20} | {'Prodi':<20} | Tahun | IPK")
    print("  " + "─" * 80)
    m1.tampilkan()
    m2.tampilkan()
    m3.tampilkan()

    # ─── Nilai default ────────────────────────────────────────────────
    demo_nilai_default()

    print("\n─── KESIMPULAN ──────────────────────────────")
    print("  Konstruktor Default     : argumen opsional/default pada __init__")
    print("  Konstruktor Non-Default : mengisi argumen spesifik")
    print("  Overloading Konstruktor : menggunakan default parameters pada Python")
    print("  self                    : digunakan untuk merujuk ke instance saat ini")
    print("  Copy Constructor        : disimulasikan dengan argumen opsional 'lain'")

if __name__ == "__main__":
    main()

